File: C++/376.cpp
__________________________________________________________________________________________________
sample 4 ms submission
class Solution {
public:
    int wiggleMaxLength(vector<int>& nums) {
        if(nums.size()<2)
            return nums.size();
        int up[nums.size()];
        int down[nums.size()];
        
        up[0] = down[0] = 1;
        for(int i=1;i<nums.size();i++){
            if(nums[i]>nums[i-1]){
                up[i] = max(up[i-1],down[i-1]+1);
                down[i] = down[i-1];
            }
            else if(nums[i]<nums[i-1]){
                down[i]=max(down[i-1],up[i-1]+1);
                up[i] = up[i-1];
            }
            else{
                down[i] = down[i-1];
                up[i] = up[i-1];
            }

        }
        return max(up[nums.size()-1],down[nums.size()-1]);
    }
};
__________________________________________________________________________________________________
sample 8444 kb submission
class Solution {
public:
    int wiggleMaxLength(vector<int>& nums) {
        // Solution 1: O(n) DP solution.
        int l = nums.size(), down = 1, up = 1; 
        if(l < 2)
            return l;
        for(int i=1; i < l; i++){
            if(nums[i] > nums[i-1]){
                up = down + 1;
            }
            else if(nums[i] < nums[i-1]){
                down = up + 1;
            } // else, up and down remain the same.
        }
        return max(up, down);
    }
};
__________________________________________________________________________________________________

File: Java/376.java
__________________________________________________________________________________________________
sample 0 ms submission
public class Solution {
    public int wiggleMaxLength(int[] nums) {
        if (nums.length < 2)
            return nums.length;
        int down = 1, up = 1;
        for (int i = 1; i < nums.length; i++) {
            if (nums[i] > nums[i - 1])
                up = down + 1;
            else if (nums[i] < nums[i - 1])
                down = up + 1;
        }
        return Math.max(down, up);
    }
}
__________________________________________________________________________________________________
sample 32900 kb submission
class Solution {
    public int wiggleMaxLength(int[] nums) {
        if (null == nums || 0 == nums.length) {
            return 0;
        }
        int max = 1;
        int[] up = new int[nums.length];
        int[] down = new int[nums.length];
        Arrays.fill(up, 1);
        Arrays.fill(down, 1);
        for (int i = 1; i < nums.length; i++) {
            for (int j = 0; j < i; j++) {
                if (nums[i] > nums[j]) {
                    up[i] = Math.max(up[i - 1], down[j] + 1);
                } else if (nums[i] < nums[j]) {
                    down[i] = Math.max(down[i - 1], up[j] + 1);
                }
            }
            max = Math.max(max, up[i]);
            max = Math.max(max, down[i]);
        }
        return max;
    }
}
__________________________________________________________________________________________________
